base/src/main/java/proguard/optimize/inline/BaseLambdaInliner.java

```diff
@@ -0,0 +1,343 @@
+import proguard.classfile.instruction.VariableInstruction;
+import proguard.classfile.instruction.visitor.AllInstructionVisitor;
+import proguard.classfile.instruction.visitor.InstructionCounter;
+import proguard.classfile.instruction.visitor.InstructionOpCodeFilter;
+import proguard.classfile.instruction.visitor.InstructionVisitor;
+import proguard.classfile.util.ClassReferenceInitializer;
+import proguard.classfile.util.ClassUtil;
+import proguard.classfile.util.InternalTypeEnumeration;
+import proguard.classfile.visitor.MemberVisitor;
+import proguard.evaluation.PartialEvaluator;
+import proguard.evaluation.TracedStack;
+import proguard.optimize.inline.lambda_locator.Lambda;
+import proguard.optimize.peephole.MethodInliner;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.lang.Math.max;
+
+public class BaseLambdaInliner implements MemberVisitor, InstructionVisitor, ConstantVisitor {
+    private final Clazz consumingClass;
+    private final Method consumingMethod;
+    private final AppView appView;
+    private final Lambda lambda;
+    private final CodeAttributeEditor codeAttributeEditor;
+    private final boolean isStatic;
+    private final int calledLambdaIndex;
+    private final int sizeAdjustedLamdaIndex;
+    private final PartialEvaluator partialEvaluator;
+    private Clazz interfaceClass;
+    private Clazz lambdaClass;
+    private Method lambdaInvokeMethod;
+    private String bridgeDescriptor;
+    private Method inlinedLambdaMethod;
+    private Method staticInvokeMethod;
+    private InterfaceMethodrefConstant referencedInterfaceConstant;
+    private final List<Integer> invokeMethodCallOffsets;
+
+    public BaseLambdaInliner(AppView appView, Clazz consumingClass, Method consumingMethod, int calledLambdaIndex, Lambda lambda) {
+        this.consumingClass = consumingClass;
+        this.consumingMethod = consumingMethod;
+        this.appView = appView;
+        this.lambda = lambda;
+        this.codeAttributeEditor = new CodeAttributeEditor();
+        this.isStatic = (consumingMethod.getAccessFlags() & AccessConstants.STATIC) != 0;
+        this.calledLambdaIndex = calledLambdaIndex + (isStatic ? 0 : 1);
+        this.sizeAdjustedLamdaIndex = ClassUtil.internalMethodVariableIndex(consumingMethod.getDescriptor(consumingClass), true, this.calledLambdaIndex);
+        this.partialEvaluator = new PartialEvaluator();
+        this.invokeMethodCallOffsets = new ArrayList<>();
+    }
+
+    /**
+     * Idea: Inline a lambda by just giving it a consuming method and class in which it has to inline the particular
+     * lambda,  a reference to the newly created function with the inlined lambda is returned, the caller can then
+     * replace the call instruction with a call to this newly created function. This way we don't need to think about
+     * how it is called exactly, it can be used as a higher abstraction in other places.
+     */
+    public Method inline() {
```

Comment:
What is needed to use the existing `MethodInliner` instead of this new implementation?

Reply:
Well this implementation makes use of `MethodInliner` but needs to do a lot more than just inlining a method. It needs to copy the method we are inlining in, it needs to do things like remove an argument from the method we are inlining in. Because we remove an argument we can't use this argument to call other methods from within the consuming method, so we need to either also recursively inline that argument in those methods or in this simplified implementation stop inlining if that is the case. The method we want to inline is also a virtual member method from the lambda class which the method inliner normally will not inline, that's why we first copy it to the consuming class and make it static. That virtual member method being the invoke method is also not actually the method we want to inline, we want to inline the invoke method called by this invoke method because this first invoke method is actually a bridge method that is type erased and does a bunch of casting. This casting reduces performance so we inline the actual invoke method that has the right argument types and does not need any casting. So I feel like this is a bit much to attempt to squeeze into the existing `MethodInliner` which only does a very basic operation that is only one tiny part of this inliner.